[FX] Fix retracing wrapped functions

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58061 [FX] Fix retracing wrapped functions**

This resolves https://github.com/pytorch/pytorch/issues/58004 by emitting calls to `wrap()` in the generated code. Some TODOs/open questions:

- [ ] Might need to handle wrapped methods as well https://github.com/pytorch/pytorch/blob/ac44569b0d87ef7c2c8c26f3057f797f40cc6111/torch/fx/symbolic_trace.py#L497
- [x] Had to do some namespace hackery to get `wrap()` accessible by FQN: https://github.com/pytorch/pytorch/pull/58061/files#diff-87b3a9c7bd0a9cc377ec096f7cfdfb20f64454f1df81adeb893b090e662ea161R724
- [ ] This potentially leaks the global namespace dict for any `GraphModule` containing wrapped functions. Is this a problem? 

Differential Revision: [D28358801](https://our.internmc.facebook.com/intern/diff/D28358801)

Commit:
[FX] Fix retracing wrapped functions

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Fix retracing wrapped functions"


This resolves https://github.com/pytorch/pytorch/issues/58004 by emitting calls to `wrap()` in the generated code. Some TODOs/open questions:

- [ ] Might need to handle wrapped methods as well https://github.com/pytorch/pytorch/blob/ac44569b0d87ef7c2c8c26f3057f797f40cc6111/torch/fx/symbolic_trace.py#L497
- [ ] Had to do some namespace hackery to get `wrap()` accessible by FQN: https://github.com/pytorch/pytorch/pull/58061/files#diff-87b3a9c7bd0a9cc377ec096f7cfdfb20f64454f1df81adeb893b090e662ea161R724
- [ ] This potentially leaks the global namespace dict for any `GraphModule` containing wrapped functions. Is this a problem? 

Differential Revision: [D28358801](https://our.internmc.facebook.com/intern/diff/D28358801)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28358801).

Commit:
Update on "[WIP][FX] Fix retracing wrapped functions"


This resolves https://github.com/pytorch/pytorch/issues/58004 by emitting calls to `wrap()` in the generated code. Some TODOs/open questions:

- [ ] Might need to handle wrapped methods as well https://github.com/pytorch/pytorch/blob/ac44569b0d87ef7c2c8c26f3057f797f40cc6111/torch/fx/symbolic_trace.py#L497
- [ ] Had to do some namespace hackery to get `wrap()` accessible by FQN: https://github.com/pytorch/pytorch/pull/58061/files#diff-87b3a9c7bd0a9cc377ec096f7cfdfb20f64454f1df81adeb893b090e662ea161R724
- [ ] This potentially leaks the global namespace dict for any `GraphModule` containing wrapped functions. Is this a problem? 

Differential Revision: [D28358801](https://our.internmc.facebook.com/intern/diff/D28358801)

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Fix retracing wrapped functions"


This resolves https://github.com/pytorch/pytorch/issues/58004 by emitting calls to `wrap()` in the generated code. Some TODOs/open questions:

- [ ] Might need to handle wrapped methods as well https://github.com/pytorch/pytorch/blob/ac44569b0d87ef7c2c8c26f3057f797f40cc6111/torch/fx/symbolic_trace.py#L497
- [ ] Had to do some namespace hackery to get `wrap()` accessible by FQN: https://github.com/pytorch/pytorch/pull/58061/files#diff-87b3a9c7bd0a9cc377ec096f7cfdfb20f64454f1df81adeb893b090e662ea161R724
- [ ] This potentially leaks the global namespace dict for any `GraphModule` containing wrapped functions. Is this a problem? 

Differential Revision: [D28358801](https://our.internmc.facebook.com/intern/diff/D28358801)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28358801).

suo commented on torch/fx/symbolic_trace.py:
```diff
@@ -717,6 +719,10 @@ def my_custom_function(x, y):
 
+# HACK: `torch.fx.symbolic_trace` refers to the function, not the module
+# wrap's default `__module__` is `torch.fx.symbolic_trace`. This leads to fuckery
+wrap.__module__ = 'torch.fx'
```
I think the real problem here is that this file (`symbolic_trace.py`) shadows the symbol that `torch/__init__.py` exports from this file (`symbolic_trace`). This is probably bad news bears in general (e.g. `import torch.fx.symbolic_trace` and `from torch.fx import symbolic_trace` will do different things!), I would suggest making this file `_symbolic_trace.py` instead. Or if you don't want it to be hidden, `symbolic_tracer.py` maybe?

Commit:
Update on "[WIP][FX] Fix retracing wrapped functions"


This resolves https://github.com/pytorch/pytorch/issues/58004 by emitting calls to `wrap()` in the generated code. Some TODOs/open questions:

- [ ] Might need to handle wrapped methods as well https://github.com/pytorch/pytorch/blob/ac44569b0d87ef7c2c8c26f3057f797f40cc6111/torch/fx/symbolic_trace.py#L497
- [ ] Had to do some namespace hackery to get `wrap()` accessible by FQN: https://github.com/pytorch/pytorch/pull/58061/files#diff-87b3a9c7bd0a9cc377ec096f7cfdfb20f64454f1df81adeb893b090e662ea161R724
- [ ] This potentially leaks the global namespace dict for any `GraphModule` containing wrapped functions. Is this a problem? 

Differential Revision: [D28358801](https://our.internmc.facebook.com/intern/diff/D28358801)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28358801).

Commit:
Update on "[WIP][FX] Fix retracing wrapped functions"


This resolves https://github.com/pytorch/pytorch/issues/58004 by emitting calls to `wrap()` in the generated code. Some TODOs/open questions:

- [ ] Might need to handle wrapped methods as well https://github.com/pytorch/pytorch/blob/ac44569b0d87ef7c2c8c26f3057f797f40cc6111/torch/fx/symbolic_trace.py#L497
- [ ] Had to do some namespace hackery to get `wrap()` accessible by FQN: https://github.com/pytorch/pytorch/pull/58061/files#diff-87b3a9c7bd0a9cc377ec096f7cfdfb20f64454f1df81adeb893b090e662ea161R724
- [ ] This potentially leaks the global namespace dict for any `GraphModule` containing wrapped functions. Is this a problem? 

Differential Revision: [D28358801](https://our.internmc.facebook.com/intern/diff/D28358801)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28358801).

suo reviewed (approved):
lgtm

Commit:
Update on "[FX] Fix retracing wrapped functions"


This resolves https://github.com/pytorch/pytorch/issues/58004 by emitting calls to `wrap()` in the generated code. Some TODOs/open questions:

- [ ] Might need to handle wrapped methods as well https://github.com/pytorch/pytorch/blob/ac44569b0d87ef7c2c8c26f3057f797f40cc6111/torch/fx/symbolic_trace.py#L497
- [x] Had to do some namespace hackery to get `wrap()` accessible by FQN: https://github.com/pytorch/pytorch/pull/58061/files#diff-87b3a9c7bd0a9cc377ec096f7cfdfb20f64454f1df81adeb893b090e662ea161R724
- [ ] This potentially leaks the global namespace dict for any `GraphModule` containing wrapped functions. Is this a problem? 

Differential Revision: [D28358801](https://our.internmc.facebook.com/intern/diff/D28358801)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28358801).
